### PR TITLE
[FSDP] Sharded Grad Scaler

### PR DESCRIPTION
**Pull request by sisilmehta2000**
Summary: Adding in a shard aware grad scaler for FSDP+MixedPrecision support

Test Plan: Tests added

Differential Revision: D35988676

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:1]: you can reference other files in test/distributed/fsdp to fix the lint here

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+params = "cpu_offload,backward_prefetch,sharding_strategy,mixed_precision"
+cpu_offload_config = [CPUOffload(offload_params=True), CPUOffload(offload_params=False)]
+backward_prefetch_config = [BackwardPrefetch.BACKWARD_PRE, BackwardPrefetch.BACKWARD_POST, None]
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:30]: In the interest of TTS, I'm pretty sure we can omit backward prefetch tests as I don't expect them to have implication for sharded grad scaling. cc @zhaojuanmao just in case I'm missing something

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        ) if mixed_precision else None
+        for fsdp_init_mode in init_modes:
+            self._test_identical_outputs(
+                NestedWrappedModule,
+                fsdp_init_mode=fsdp_init_mode,
+                cpu_offload=cpu_offload,
+                backward_prefetch=backward_prefetch,
+                sharding_strategy=sharding_strategy,
+                mixed_precision=mp,
+                enable_sharded_grad_scaler=True,
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:86]: is there a reason we're testing sharded grad scaler when mixed precision is _not_ enabled?

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+def _refresh_per_optimizer_state():
+    return {"stage": OptState.READY, "found_inf_per_device": {}}
+
+
+def _is_supported_device(tensor: torch.Tensor):
+    return tensor.is_cuda or tensor.device.type in ("xla", "cpu")
+
+
+class _GeneralMultiDeviceReplicator(_MultiDeviceReplicator):
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:21]: Add a docstring about what this class enables over `_MultiDeviceReplicator` to explain why we need it? Is it just to override the supported devices?

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+import torch.distributed as dist
+from torch.optim.sgd import SGD
+
+
+def _refresh_per_optimizer_state():
+    return {"stage": OptState.READY, "found_inf_per_device": {}}
+
+
+def _is_supported_device(tensor: torch.Tensor):
+    return tensor.is_cuda or tensor.device.type in ("xla", "cpu")
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:18]: FSDP doesn't support XLA at the moment so we can possibly omit, though it might just be worth keeping for when we eventually do support XLA, I'll let you make the call

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+class _GeneralMultiDeviceReplicator(_MultiDeviceReplicator):
+    def __init__(self, master_tensor: torch.Tensor) -> None:
+        assert _is_supported_device(master_tensor)
+        self.master = master_tensor
+        self._per_device_tensors: Dict[torch.device, torch.Tensor] = {}
+
+
+class ShardedGradScaler(GradScaler):
+    """
+    ShardedGradScaler helps performance gradient scaling in a shard aware manner. It extends
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:30]: nit: helps perform

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+        def apply_scale(val: Union[torch.Tensor, abc.Iterable[torch.Tensor]]) -> Union[torch.Tensor, abc.Iterable[torch.Tensor]]:
+            if isinstance(val, torch.Tensor):
+                assert _is_supported_device(val)
+                if len(stash) == 0:
+                    if self._scale is None:
+                        self._lazy_init_scale_growth_tracker(val.device)
+                    assert self._scale is not None
+                stash.append(_GeneralMultiDeviceReplicator(self._scale))
+                scaled_val = val * stash[0].get(val.device)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:120]: Trying to understand, for lists of tensors input, we can append to stash many times, but we always use stash[0] still?

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+        assert inv_scale.numel() == 1, "inv_scale must be a 1-element tensor."
+        assert found_inf.numel() == 1, "found_inf must be a 1-element tensor."
+
+        expected_device = grads[0].device
+        for grad in grads:
+            for tensor in grad:
+                try:
+                    assert tensor.device == expected_device, "grads must be on the same device"
+                except AssertionError:
+                    logging.error("tensor device is %s and expected device is %s" % (tensor.device, expected_device))
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:150]: seems a bit roundabout to try, raise an assert, and then catch the assert to warn? Could we just warn directly?

But also, should we just throw here because grads on different device than params seems unexpected, it would have an issue in `opt.step()`?

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+                # check for non_overlapping_and_dense doesn't exist in the python world
+                # as remarked here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/AmpKernels.cu#L108
+                # we assume tensor is not MTA(multi tensor apply) safe. iterate through each item regardless of dtype
+                # TODO faster logic?
+                if torch.isinf(tensor).any().item() is True or torch.isnan(tensor).any().item() is True:  # type: ignore
+                    found_inf.data = torch.tensor([1.0])
+                    break
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:158]: this means some grads might be scaled and some might not be? I guess that is fine, because when we find inf, opt.step() will be a noop, and eventually we just won't use these grads basically right?

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+            if v.device.type == "cpu":
+                v_on_cuda = v.cuda()
+                last_handle = dist.all_reduce(v_on_cuda, async_op=True, group=self.process_group)
+                v_on_cuda.cpu()
+            else:
+                last_handle = dist.all_reduce(v, async_op=True, group=self.process_group)
+
+        # Make sure that the calls are done before moving out.
+        if last_handle:
+            last_handle.wait()
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:247]: hmm, i guess this is sort of using an assumption of NCCL IIUC, i.e. if we submit comm ops on the same stream on the same nccl pg (which is what we do here) then nccl takes care to run those sequentially currently I believe, even if they are not data dependent ops. But I guess to be more backend agnostic or robust, it might be worth to collect all the futures (shouldn't be too many as they are per device) and wait them all with `torch.futures.wait_all`

but this can be done as a follow up, I think currently this should be fine

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+
+        # To set up _amp_foreach_non_finite_check_and_unscale_, split grads by device and dtype.
+        # There could be thousands of grads, so we'd like to iterate through them just once.
+        # However, we don't know their devices or dtypes in advance.
+
+        # https://stackoverflow.com/questions/5029934/defaultdict-of-defaultdict
+        # Google says mypy struggles with defaultdicts type annotations.
+        per_device_and_dtype_grads = defaultdict(lambda: defaultdict(list))  # type: ignore[var-annotated]
+        with torch.no_grad():
+            for group in optimizer.param_groups:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:176]: When FSDP supports optimizer param groups, we should add the relevant tests to sharded grad scaler as well.

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+                stash.append(_GeneralMultiDeviceReplicator(self._scale))
+                scaled_val = val * stash[0].get(val.device)
+                # Here we ensure the return dtype is the same as the outputs dtype.
+                # For the FSDP + Mixed Precision use case, the loss output is in the Mixed Precision
+                # format (fp16, bf16) and so the scaled loss should be of the same dtype.
+                return scaled_val.type(val.dtype)
+            elif isinstance(val, abc.Iterable):
+                iterable = map(apply_scale, val)
+                if isinstance(val, (list, tuple)):
+                    return type(val)(iterable)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:128]: Can we have tests where a loss that is a list of tensors is scaled?

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+        optimizer_state["stage"] = OptState.UNSCALED
+
+        # Synchronize the detected inf across the ranks
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+        last_handle = None
+
+        for v in optimizer_state["found_inf_per_device"].values():
+            if v.device.type == "cpu":
+                v_on_cuda = v.cuda()
+                last_handle = dist.all_reduce(v_on_cuda, async_op=True, group=self.process_group)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:240]: Not sure if I completely follow this or I'm missing something, but doesn't this _not_ update the current `v` in optimizer_state[found_inf_per_device]?

v_on_cuda should be an entirely different tensor than v, and we pass this into all_reduce. Then we do v_on_cuda.cpu(), but I'm not sure what that does, it creates a CPU tensor that we don't assign?

Trying to figure out if the intention here is to convert the current `v` inplace and make sure `v` updates it, but if that is the case based on my understanding this might not be happening.

**Review by rohan-varma — COMMENTED** [review]
Looks great to me overall, thanks for working on this! Have a few comments / questions, and will also wait for the additional tests you mentioned as TODO. Thanks for putting this together so quickly!

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py (hunk 3 above)
Comment (sisilmehta2000) [reply test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:86]: Yeah, I was using it as a baseline test to compare DDP vs FSDP (sharded grad scaler turned on and off). That way the test actually compares the params after the iterations. This was pretty useful the first few times I ran it to identify edge cases where there were bugs in the sharded grad scaler code.

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 4 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:21]: Correct just to override the supported devices. Leaving comment here.

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 7 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:120]: indent bug. fixed.

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 8 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:150]: Yeah raising error here

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 9 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:158]: Yup correct. Once we find inf/nan, opt.step is skipped.

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 11 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:176]: Leaving a TODO here for that.

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 13 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:240]: Good catch. Fixed.

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 10 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:247]: done.

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py (hunk 3 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:86]: Sounds good!

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 12 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py]: added

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_utils import (
+    TestCase, run_tests,
+    instantiate_parametrized_tests,
+    parametrize,
+    TEST_WITH_DEV_DBG_ASAN,
+)
+
+
+if not dist.is_available():
```
Comment (sisilmehta2000) [inline test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:22]: Need to move this check into the ditributed tests on line 115

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    str(CPUOffload(offload_params=True)): "offload_true",
+    str(CPUOffload(offload_params=False)): "offload_false",
+    str(ShardingStrategy.SHARD_GRAD_OP): "shard_grad_op",
+    "enable_mixed_precision": "mixed_precision"
+}
+
+subtest_name = functools.partial(subtest_name, test_name_mapping)
+
+
+class TestShardGradScaler(TestCase):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:53]: These tests don't really seem to depend on shards / multiple ranks? Should we file an issue to test this against traditional grad scaler and verify we have parity?

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        return modes
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(params, configs, subtest_name)
+    def test_scaler_enabled(self, cpu_offload, sharding_strategy, mixed_precision):
+        init_modes = self._get_init_modes_for_test(cpu_offload)
+        mp = MixedPrecision(
+            param_dtype=torch.float16,
+            reduce_dtype=torch.float16,
+            buffer_dtype=torch.float16,
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:137]: As a follow up, can we also add bfloat16 tests. bfloat16 is hardware conditional, there are some logic in test_fsdp_mixed_precision to make sure we only run it where appropriate.

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks a bunch for working on this and thank you for the excellent testing coverage!

Please fix typing and CI issues before landing. Also had some comments / suggestions around better testing, but feel like they can be addressed as follow up tasks (if you decide not to address them in this PR, can you file follow up issues?)

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+        growth_interval (int, optional, default=2000):  Number of consecutive iterations without inf/NaN gradients
+            that must occur for the scale to be multiplied by ``growth_factor``.
+        enabled (bool, optional, default=True):  If ``False``, disables gradient scaling. :meth:`step` simply
+            invokes the underlying ``optimizer.step()``, and other methods become no-ops.
+        process_group (ProcessGroup, optional, default=torch.distributed.group.WORLD):
+            process group for sharding
+    """
+    def __init__(
+        self,
+        init_scale: float = 2.0 ** 16,
```
Comment (rohan-varma) [inline torch/distributed/fsdp/sharded_grad_scaler.py:81]: Do we feel like these defaults are well-tuned or expect the user have to tune them for their use case?

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py (hunk 15 above)
Comment (sisilmehta2000) [reply test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:53]: They dont directly depend on shard/multiple ranks. GradScaler.py has a check where if cuda is not available, it disables GradScaling. https://github.com/pytorch/pytorch/blob/bbc3cc6718d841665ab2677477d458ed97b75609/torch/cuda/amp/grad_scaler.py#L116

File: torch/distributed/fsdp/sharded_grad_scaler.py (hunk 17 above)
Comment (sisilmehta2000) [reply torch/distributed/fsdp/sharded_grad_scaler.py:81]: I used the ones from the Fairscale implementation. I think will need a little experiment across different types of models to change the defaults.

File: test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py (hunk 16 above)
Comment (sisilmehta2000) [reply test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:140]: bfloat16 is tested in test_mixed_precision.py above.
